apps/opencs/view/render/pathgrid.cpp

```diff
@@ -0,0 +1,547 @@
+                }
+            }
+
+            CSMWorld::IdTree* model = dynamic_cast<CSMWorld::IdTree*>(mData.getTableModel(
+                CSMWorld::UniversalId::Type_Pathgrids));
+
+            int parentColumn = mPathgridCollection.findColumnIndex(CSMWorld::Columns::ColumnId_PathgridEdges);
+
+            std::set<int, std::greater<int> >::iterator row;
+            for (row = rowsToRemove.begin(); row != rowsToRemove.end(); ++row)
+            {
+                commands.push(new CSMWorld::DeleteNestedCommand(*model, mId, *row, parentColumn));
+            }
+        }
+    }
+
+    osg::ref_ptr<PathgridTag> Pathgrid::getTag() const
+    {
+        return mTag;
+    }
+
+    void Pathgrid::recreateGeometry()
+    {
+        mChangeGeometry = true;
+    }
+
+    void Pathgrid::removeGeometry()
+    {
+        mRemoveGeometry = true;
+
+
+    }
+
+    void Pathgrid::update()
+    {
+        if (mRemoveGeometry)
+        {
+            removePathgridGeometry();
+            removeSelectedGeometry();
+        }
+        else if (mChangeGeometry)
+        {
+            createGeometry();
+        }
+
+        mChangeGeometry = false;
+        mRemoveGeometry = false;
+    }
+
+    void Pathgrid::createGeometry()
+    {
+        const CSMWorld::Pathgrid* source = getPathgridSource();
+        if (source)
+        {
+            removePathgridGeometry();
+            mPathgridGeometry = SceneUtil::createPathgridGeometry(*source);
+            mPathgridGeode->addDrawable(mPathgridGeometry);
+
+            createSelectedGeometry(*source);
+        }
+        else
+        {
+            removeGeometry();
+        }
+    }
+
+    void Pathgrid::createSelectedGeometry()
+    {
+        const CSMWorld::Pathgrid* source = getPathgridSource();
+        if (source)
+        {
+            createSelectedGeometry(*source);
+        }
+        else
+        {
+            removeSelectedGeometry();
+        }
+    }
+
+    void Pathgrid::createSelectedGeometry(const CSMWorld::Pathgrid& source)
+    {
+        removeSelectedGeometry();
+
+        if (mConnectionIndicator)
+        {
+            NodeList tempList = NodeList(mSelected);
+
+            NodeList::iterator searchResult = std::find(tempList.begin(), tempList.end(), mConnectionNode);
+            if (searchResult != tempList.end())
+                tempList.erase(searchResult);
+
+            tempList.push_back(mConnectionNode);
+
+            mSelectedGeometry = SceneUtil::createPathgridSelectedWireframe(source, tempList);
+            mSelectedGeode->addDrawable(mSelectedGeometry);
+        }
+        else
+        {
+            mSelectedGeometry = SceneUtil::createPathgridSelectedWireframe(source, mSelected);
+            mSelectedGeode->addDrawable(mSelectedGeometry);
+        }
+    }
+
+    void Pathgrid::removePathgridGeometry()
+    {
+        if (mPathgridGeometry)
+        {
+            mPathgridGeode->removeDrawable(mPathgridGeometry);
+            mPathgridGeometry.release();
```

Comment:
Don't use release() to free a ref_ptr. That unlinks the contained object without actually deleting it, resulting in a memory leak.